iavl/iavl_tree.go

```diff
@@ -253,70 +515,253 @@ func (ndb *nodeDB) GetNode(t *IAVLTree, hash []byte) *IAVLNode {
 func (ndb *nodeDB) SaveNode(t *IAVLTree, node *IAVLNode) {
 	ndb.mtx.Lock()
 	defer ndb.mtx.Unlock()
+
 	if node.hash == nil {
-		PanicSanity("Expected to find node.hash, but none found.")
+		cmn.PanicSanity("Expected to find node.hash, but none found.")
 	}
 	if node.persisted {
-		PanicSanity("Shouldn't be calling save on an already persisted node.")
+		cmn.PanicSanity("Shouldn't be calling save on an already persisted node.")
 	}
-	/*if _, ok := ndb.cache[string(node.hash)]; ok {
-		panic("Shouldn't be calling save on an already cached node.")
-	}*/
+
 	// Save node bytes to db
 	buf := bytes.NewBuffer(nil)
 	_, err := node.writePersistBytes(t, buf)
 	if err != nil {
-		PanicCrisis(err)
+		cmn.PanicCrisis(err)
 	}
 	ndb.batch.Set(node.hash, buf.Bytes())
 	node.persisted = true
 	ndb.cacheNode(node)
-	// Re-creating the orphan,
-	// Do not garbage collect.
-	delete(ndb.orphans, string(node.hash))
-	delete(ndb.orphansPrev, string(node.hash))
 }
 
+func (ndb *nodeDB) GetRoots(t *IAVLTree) {
+	buf := ndb.db.Get(rootsKey)
+	if len(buf) == 0 {
+		cmn.PanicSanity(cmn.Fmt("Value missing for key %X", rootsKey))
+	}
+
+	//fmt.Printf("Have Versions: %X\n", buf)
+	count := int(wire.GetInt16(buf))
+	buf = buf[2:]
+
+	for i := 0; i < count; i++ {
+		bytes, n, err := wire.GetByteSlice(buf)
+		if err != nil {
+			cmn.PanicSanity(err)
+		}
+		buf = buf[n:]
+
+		//fmt.Printf("Have Node: %X\n", bytes)
+		SetValue(t.roots, i, t.ndb.GetNode(t, bytes))
+	}
+}
+
+func (ndb *nodeDB) SaveRoots(t *IAVLTree) {
+	buf, n, err := bytes.NewBuffer(nil), new(int), new(error)
+
+	wire.WriteInt16(int16(t.roots.Len()), buf, n, err)
+	for e := t.roots.Front(); e != nil; e = e.Next() {
+		value := e.Value.(*IAVLNode)
+		wire.WriteByteSlice(value.hash, buf, n, err)
+	}
+	ndb.batch.Set(rootsKey, buf.Bytes())
+}
+
+func (ndb *nodeDB) GetVersion(t *IAVLTree) {
+	buf := ndb.db.Get(versionKey)
+	if len(buf) == 0 {
+		t.version = 0
+	}
+	version, _, err := wire.GetVarint(buf)
+	if err != nil {
+		cmn.PanicSanity(err)
+	}
+	t.version = version
+}
+
+func (ndb *nodeDB) SaveVersion(t *IAVLTree) {
+	buf, n, err := bytes.NewBuffer(nil), new(int), new(error)
+
+	wire.WriteVarint(t.version, buf, n, err)
+	ndb.batch.Set(versionKey, buf.Bytes())
+}
+
+// GetOrphans from the database, their key is the root hash
+func (ndb *nodeDB) GetOrphans(hash []byte) [][]byte {
+	key := orphansKey
```

Comment:
Sure. The original style was quite explicit, but I usually prefer to encapsulate more heavily. This isn't a great example, but we should have a discussion at the retreat on style issues (they often differ quite a bit between systems and application programming).